### PR TITLE
Pimp the UiDispatcher

### PR DESCRIPTION
**Pull request by MDoerner**
This PR adds the capability to the `UIDispatcher` to start tasks on the UI thread. With this capability we can now wait for an action to finish executing on the UI thread and get the return value if a `Func<T>` has been passed in.

Moreover, I commented out the shutdown code in the `UiDispatcher` in order to avoid the exception on reload we experience so far. (Ref #3645) This does avoid said exception but the state RD reloads in is not completely clean. This is probably connected to our managed memory leak at shutdown/unload. (Ref #3675)

### CODE REVIEW

File: RetailCoder.VBE/UI/Command/MenuItems/UiDispatcher.cs [hunk 1]
```diff
@@ -55,18 +58,62 @@ public static void Invoke(Action action)
             }
         }
 
+        /// <summary>
+        /// Starts a task on the ui thread.
+        /// </summary>
+        /// <param name="action">The action that will be executed on the UI
+        /// thread.</param>
+        /// <param name="token">Optional cancellation token</param>
+        /// <param name="options">Oprional TaskCreationOptions</param>
```
Comment (Vogel612) [inline RetailCoder.VBE/UI/Command/MenuItems/UiDispatcher.cs]: typo: optional

**Comment by WaynePhillipsEA** [comment]
referencing #3645, which this PR technically closes, but we may want to leave open until the managed memory leaks have been plugged, so that the shutdown code can then hopefully be reinstated.

**Comment by bclothier** [comment]
I have a question about the `Initialize` process - it seems that we do this as a way to ensure that the `Initialize` is called on the UI thread and then we run `CheckInitialization` in all other methods. I'm wondering if it'd be more cleaner to simply name the UI thread within the `Extension` so we can always pull it out of the `Threads` collection in the ctor or perhaps store it in a static class, so that this can be always "initialized"?

**Comment by MDoerner** [comment]
The concept .Net uses for synchronization is the `SynchronisationContext`, which is not identical to a thread. The Winforms `SynchronisationContext` corresponds exactly the UI thread, other types correspond to multiple threads.

In `Extension` only the default `SynchronisationContext` exists. You can see this because, by convention, `SynchronisationContext.Current` returns `null`. This default context corresponds to the current thread and the thread pool. So, we really do not want that context. Instead we call `Initialize` only at a point where the actual UI context has been generated.

**Comment by rubberduck203** [comment]
Overall the code coverage has remained the same, but I noticed [ReferenceResolverRunnerBase dropped significantly](https://codecov.io/gh/rubberduck-vba/Rubberduck/pull/3780/diff?src=pr&el=tree#diff-UnViYmVyZHVjay5QYXJzaW5nL1ZCQS9SZWZlcmVuY2VSZXNvbHZlUnVubmVyQmFzZS5jcw==).

**Comment by MDoerner** [comment]
@rubberduck203 That file is not even in the diff of this PR.

**Pull request merged**